File: examples/node/dynamodb-query.js
// Consulta items de DynamoDB por clave de partición con Query (más eficiente que Scan).
// Uso: node dynamodb-query.js <tabla> <id>
const AWS = require('aws-sdk');

const dynamodb = new AWS.DynamoDB({
  endpoint: 'http://localhost:4566',
  region: 'us-east-1',
  credentials: { accessKeyId: 'test', secretAccessKey: 'test' },
});

const [, , tableName, id] = process.argv;

async function main() {
  if (!tableName || !id) {
    throw new Error('Uso: node dynamodb-query.js <tabla> <id>');
  }
  // Query exige una condición de igualdad sobre la clave de partición (KeyConditionExpression).
  // Con una tabla de clave simple como esta, Query devuelve como máximo un item: su verdadera
  // ventaja aparece con una clave de ordenación o un índice secundario (GSI, ver Módulo 4),
  // donde puedes acotar un rango de items relacionados sin recorrer la tabla entera.
  const result = await dynamodb.query({
    TableName: tableName,
    KeyConditionExpression: 'id = :id',
    ExpressionAttributeValues: { ':id': { S: id } },
  }).promise();

  if (!result.Items.length) {
    console.log(`No existe ningún item con id="${id}" en ${tableName}.`);
    return;
  }
  console.log(`${result.Count} item(s) encontrado(s) (escaneados: ${result.ScannedCount}):`);
  console.log(JSON.stringify(result.Items, null, 2));
}

main().catch((error) => {
  console.error('Error consultando la tabla:', error.message);
  process.exit(1);
});

File: examples/python/dynamodb_query.py
"""Consulta items de DynamoDB por clave de partición con Query (más eficiente que Scan).

Uso: python3 dynamodb_query.py <tabla> <id>
"""
import json
import sys

import boto3

dynamodb = boto3.client(
    "dynamodb",
    endpoint_url="http://localhost:4566",
    aws_access_key_id="test",
    aws_secret_access_key="test",
    region_name="us-east-1",
)


def main():
    if len(sys.argv) < 3:
        raise SystemExit("Uso: python3 dynamodb_query.py <tabla> <id>")
    table_name, item_id = sys.argv[1], sys.argv[2]

    # Query exige una condición de igualdad sobre la clave de partición (KeyConditionExpression).
    # Con una tabla de clave simple como esta, Query devuelve como máximo un item: su verdadera
    # ventaja aparece con una clave de ordenación o un índice secundario (GSI, ver Módulo 4),
    # donde puedes acotar un rango de items relacionados sin recorrer la tabla entera.
    result = dynamodb.query(
        TableName=table_name,
        KeyConditionExpression="id = :id",
        ExpressionAttributeValues={":id": {"S": item_id}},
    )

    if not result["Items"]:
        print(f'No existe ningún item con id="{item_id}" en {table_name}.')
        return
    print(f'{result["Count"]} item(s) encontrado(s) (escaneados: {result["ScannedCount"]}):')
    print(json.dumps(result["Items"], indent=2, ensure_ascii=False))


if __name__ == "__main__":
    main()
